contents/regex/Regex.md

```diff
@@ -0,0 +1,316 @@
+
+
+For a full list of possible operators, refer to this quick start guide:
+[Quick Start Guide](http://www.regular-expressions.info/quickstart.html)<br>
+
+## Search and Replace
+
+Let's take a look at some simple examples without code. For this guide, you can use any text editor that supports Regex functionality (e.g. [Notepad++](https://notepad-plus-plus.org/download/v7.3.1.html), [Sublimetext](https://www.sublimetext.com/)).
+
+Say you have a passage of text which contains NRICs to be censored.
+
+### Example
+
+> John - S1234567A <br>
+> Mary - S8192853B <br>
+> David - S1235326C
+
+This can't be done with a simple search and replace, since NRICs have different digits, and we can't specify which ones to look for. However, we know that an NRIC consists of an S, 7 digits from 0-9 and ends with a letter. Knowing this we can construct a Reges:
+
+* Find: `S[\d]{7}[A-Z]`
+* Replace: `SXXXXXXXX`
+
+As we can see, the `S` matches the very first S, [\d]{7} matches a string of 7 digits, and [A-Z] matches a single capital letter. We then replace all the critical information with a string of Xs.
+
+### Result
+
+> John - SXXXXXXXX <br>
+> Mary - SXXXXXXXX <br>
+> David - SXXXXXXXX
+
+## Capturing groups
+
+Now, imagine you have a section of text with many dates written in the US Date Format (MM/DD/YYYY).
+
+> Calendar <br>
+> 02/20/2017 Paul's Birthday <br>
+> 02/24/2017 Presentation at work <br>
+> 02/25/2017 Workshop
+
+However, you want them to be written in (DD/MM/YYYY) instead. As we can see, this isn't as simple as just finding and replacing text, we must also maintain the data in the search string (e.g. `02/20` becomes `20/02`). This can be done using a **capturing group**. A capturing group is a part of a Regex surrounded by brackets `()`. They are numbered `1`, `2`, `3`, from left to right and are used to hold captured data. For example:
+
+`(\d\d)/(\d\d)/(\d\d\d\d)` against `02/20/2017` will result in group 1 : `02`, group 2 : `20`, and group 3 : `2017`.
+
+Using this, we can construct our regex as follows:
+
+* Find: `(\d\d)/(\d\d)/(\d\d\d\d)`
+* Replace: `\2/\1/\3`
+
+### Result
+
+> Calendar <br>
+> 20/02/2017 Paul's Birthday <br>
+> 24/02/2017 Presentation at work <br>
+> 25/02/2017 Workshop
+
+-More examples-
+
+## Helpful links
+
+[Try it out online](https://regex101.com/)<br>
+[Regex Examples](http://www.regular-expressions.info/examples.html)<br>
+[Regex Course](https://regexone.com/)
+
+# Advanced topics
+
+## The Regex Engine
+
+So, how does Regex work exactly?
+
+Regex works similarly to most brute force searches. A simplified Regex engine would look something like this:
+
+1. Start from the start of the text and the start of the regex.
+2. Check if the first character matches.
+3. If it does check the second character of both strings and so on.
+4. If any character doesn't match, start from the second character of the text.
+5. Return when a match is found, or if no match is found.
```

Comment:
Use `1.` for all numbers; GFM automatically translates it to proper list numbering.